File: src/main/java/com/cout970/computer2/network/MessageButtonClick.java
package com.cout970.computer2.network;

import com.cout970.computer.network.IButtonListener;
import io.netty.buffer.ByteBuf;
import net.darkaqua.blacksmith.api.common.network.ExtendedByteBuf;
import net.darkaqua.blacksmith.api.common.network.INetworkContext;
import net.darkaqua.blacksmith.api.common.network.INetworkMessage;
import net.darkaqua.blacksmith.api.common.network.INetworkMessageHandler;
import net.darkaqua.blacksmith.api.common.tileentity.ITileEntity;
import net.darkaqua.blacksmith.api.common.util.ObjectScanner;
import net.darkaqua.blacksmith.api.common.util.vectors.Vect3i;
import net.darkaqua.blacksmith.api.common.world.IWorld;

/**
 * Created by cout970 on 07/02/2016.
 */
public class MessageButtonClick implements INetworkMessage, INetworkMessageHandler<MessageButtonClick, INetworkMessage> {

    private int buttonId;
    private int dimension;
    private Vect3i pos;

    public MessageButtonClick() {}

    public MessageButtonClick(int buttonId, int dimension, Vect3i pos) {
        this.buttonId = buttonId;
        this.dimension = dimension;
        this.pos = pos;
    }

    @Override
    public void fromBytes(ByteBuf buf, ExtendedByteBuf helper) {
        buttonId = buf.readInt();
        dimension = buf.readInt();
        pos = helper.readVect3i();
    }

    @Override
    public void toBytes(ByteBuf buf, ExtendedByteBuf helper) {
        buf.writeInt(buttonId);
        buf.writeInt(dimension);
        helper.writeVect3i(pos);
    }

    @Override
    public INetworkMessage onMessage(MessageButtonClick message, INetworkContext context) {
        context.getServerContext().addScheduledTask(new Runnable() {
            @Override
            public void run() {
                IWorld world = context.getServerContext().getServer().getWorld(message.dimension);
                ITileEntity tile = world.getTileEntity(message.pos);
                IButtonListener listener = ObjectScanner.findInTileEntity(tile, IButtonListener.class);
                listener.onPress(message.buttonId);
            }
        });
        return null;
    }
}

File: src/main/kotlin/com/cout970/computer/network/MessageButtonClick.kt
package com.cout970.computer.network

import com.cout970.computer.util.readBlockPos
import com.cout970.computer.util.writeBlockPos
import io.netty.buffer.ByteBuf
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.common.network.simpleimpl.IMessage
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext

/**
 * Created by cout970 on 20/05/2016.
 */
class MessageButtonClick() : IMessage, IMessageHandler<MessageButtonClick, IMessage>{

    var buttonId: Int = 0
    var dimension: Int = 0
    var pos: BlockPos? = null

    constructor(buttonId : Int, dimension : Int, pos: BlockPos): this(){
        this.buttonId = buttonId
        this.dimension = dimension
        this.pos = pos
    }

    override fun fromBytes(buf: ByteBuf?) {
        if(buf == null)return
        buttonId = buf.readInt();
        dimension = buf.readInt()
        pos = buf.readBlockPos()
    }

    override fun toBytes(buf: ByteBuf?) {
        if(buf == null)return
        buf.writeInt(buttonId)
        buf.writeInt(dimension)
        buf.writeBlockPos(pos!!)
    }

    override fun onMessage(message: MessageButtonClick?, ctx: MessageContext?): IMessage? {
        val context = ctx!!.serverHandler
        context.playerEntity.mcServer.addScheduledTask({
            val world = context.playerEntity.mcServer.worldServerForDimension(message!!.dimension)
            val tile = world.getTileEntity(message.pos)
            val listener = tile as? IServerButtonListener
            listener?.onPress(message.buttonId)
        })
        return null
    }
}